#' Call `ggtree()` and add metadata
#'
#' This function calls [ggtree::ggtree()] on the tree component of a tidygenomes object and
#' adds all available genome and node metadata.
#'
#' @param tg A tidygenomes object
#' @param ... Extra arguments to pass on to `ggtree()`
#' 
#' @return A ggplot object
#' 
#' @importFrom ggtree %<+%
#' @export
ggtree_augmented <- function(tg, ...) {
  
  if (! is.null(tg$phylogroups)) {
    
    tg <- tg %>% modify_at("nodes", left_join, tg$phylogroups, by = "phylogroup")
    
  }
  
  if (! is.null(tg$patterns)) {
    
    tg <- tg %>% modify_at("nodes", left_join, tg$patterns, by = "node")
    
  }
  
  nodes <- 
    tg$nodes %>%
    left_join(tg$genomes, by = "node") %>%
    select(label = node, everything())
  
  ggtree::ggtree(tg$tree, ...) %<+%
    nodes 
  
}

#' Construct an "upset plot" to explore gene presence/absence
#'
#' This function constructs an "upset plot" showing patterns of gene familiy
#' presence/absence in order of pattern frequency (number of gene families
#' following the pattern). The genomes are ordered along their phylogeny.
#'
#' @param tg A tidygenomes object
#' @param genome_name Name of a variable (unquoted) from the genome or
#'   phylogroup table that should be used to name the genomes
#' @param genome_col Name of a variable (unquoted) from the genome or phylogroup
#'   table that should be used to color the genome names
#' @param genome_bold Name of a logical variable (unquoted) from the genome or
#'   phylogroup table that indicates which genome names should be in bold
#' @param color_scale Ggplot color scale generated by a scale_color_... type
#'   function
#' @param n The number of presence/absence patterns to plot
#' 
#' @return A ggplot object
#' 
#' @export
upset_plot <- function(
  tg, genome_name, genome_col, genome_bold, 
  color_scale = scale_color_brewer(palette = "Paired", guide = "none"), 
  n = 50) {
  
  genome_name <- rlang::enexpr(genome_name) 
  genome_col <- rlang::enexpr(genome_col)
  genome_bold <- rlang::enexpr(genome_bold)
  
  if (is.null(tg$patterns)) stop("no patterns found")
  
  tg$patterns <-
    tg$patterns %>%
    arrange(desc(frequency)) %>%
    slice(1:UQ(n)) %>%
    mutate(pattern_fct = factor(pattern, levels = pattern)) 
  
  genomes <- 
    genomes_extended(tg) %>%
    mutate(node_fct = factor(
      node, levels = !! tipnodes_ladderized(tg$tree)
    )) %>%
    arrange(desc(node_fct)) %>%
    mutate(genome_name_fct = factor(
      !! genome_name, levels = !! genome_name
    )) %>%
    rename(genome_bold = !! genome_bold)
  
  theme_upset <- 
    theme(
      axis.text.x = element_blank(), 
      axis.title = element_blank(),
      axis.ticks = element_blank(), 
      panel.background = element_rect(fill = "transparent"), 
      plot.background = element_rect(fill = "transparent")
    )
  
  plot_main <- 
    tg$components %>%
    right_join(tg$patterns, by = "pattern") %>%
    left_join(genomes, by = "genome") %>%
    ggplot(aes(x = pattern_fct, y = genome_name_fct, group = pattern_fct)) +
    geom_line(col = "grey") +
    geom_point(size = 3, aes(col = !! genome_col)) +
    color_scale +
    theme_bw() +
    theme_upset +
    theme(axis.text.y = element_text(face = if_else(genomes$genome_bold, "bold", "plain")))
  
  plot_marg <- 
    tg$patterns %>% 
    ggplot(aes(x = pattern_fct, y = frequency)) +
    geom_histogram(stat = "identity") +
    # geom_text(aes(label = n), vjust= - 0.25, size = 2) + 
    ylim(c(0, 1.1 * max(tg$patterns$frequency))) +
    theme_bw() +
    theme_upset +
    theme(
      panel.border = element_blank(),
      panel.grid = element_blank()
    )
  
  egg::ggarrange(
    plot_marg, plot_main,
    nrow = 2, ncol = 1, heights = c(1, nrow(genomes) / 50)
  )
  
}